app/models/service/resource_linking.rb

```diff
@@ -0,0 +1,28 @@
+module Service::ResourceLinking
+  extend ActiveSupport::Concern
+
+  def add_provider_vms(provider, uid_ems_array)
+    vm_id_array = Array(uid_ems_array).compact.uniq
```

Comment:
Minor, but the name `vm_id_array` is misleading as it is the `uid` that is expected here.